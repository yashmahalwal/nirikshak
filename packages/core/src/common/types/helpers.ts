import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
