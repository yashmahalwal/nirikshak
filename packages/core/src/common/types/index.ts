export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
