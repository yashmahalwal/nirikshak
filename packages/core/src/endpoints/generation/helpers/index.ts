export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
