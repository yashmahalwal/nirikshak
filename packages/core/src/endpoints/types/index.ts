export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
