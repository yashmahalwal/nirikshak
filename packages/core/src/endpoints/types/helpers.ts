import { ResourceString, isResourceString } from "./resourceString";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { Primitives } from "../../common/types/helpers";

export type MethodType = "DELETE" | "GET" | "PUT" | "POST" | "PATCH";
export function isMethodType(key: any): key is MethodType {
    if (typeof key !== "string") return false;

    switch (key) {
        case "DELETE":
        case "GET":
        case "PUT":
        case "POST":
        case "PATCH":
            return true;
        default:
            return false;
    }
}

export type Cases = "NEGATIVE" | "POSITIVE" | "DESTRUCTIVE";

export function isCaseType(key: any): key is Cases {
    if (typeof key !== "string") return false;
    switch (key) {
        case "NEGATIVE":
        case "POSITIVE":
        case "DESTRUCTIVE":
            return true;
        default:
            return false;
    }
}

export type BaseType =
    | FakerType
    | CustomFunctionType
    | ResourceString
    | Literal;

export function isBaseType(input: any): input is BaseType {
    return (
        isFakerType(input) ||
        isCustomFunction(input) ||
        isResourceString(input) ||
        isLiteral(input)
    );
}

export type Base = BaseType | Base[];

export function isBase(input: any): input is Base {
    if (Array.isArray(input)) return input.every((value) => isBase(value));
    return isBaseType(input);
}

export type WithModifiers<T extends Base | BodyType> = (T extends Base
    ? { type: Base }
    : { field: BodyType }) & {
    plural?: boolean;
    nullable?: boolean;
    optional?: boolean;
};

// Checking the validity of the modifier fields of the HOT
function checkModifiers(input: WithModifiers<Base | BodyType>): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersBodyType(
    input: any
): input is WithModifiers<BodyType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isBodyType(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBase(input: any): input is WithModifiers<Base> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBase(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<Base | BodyType> {
    return isWithModifiersBase(input) || isWithModifiersBodyType(input);
}

export type OneOfEntries =
    | { fields: (BodyType | WithModifiers<BodyType>)[] }
    | { types: (Base | WithModifiers<Base>)[] }
    | {
          fields: (BodyType | WithModifiers<BodyType>)[];
          types: (Base | WithModifiers<Base>)[];
      };

// type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBase(entry) || isWithModifiersBase(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) => isBodyType(entry) || isWithModifiersBodyType(entry)
              )
            : true);

    return value;
}

export interface BodyType {
    [key: string]:
        | ResourceString
        | Exclude<Base, ResourceString>
        | WithModifiers<Base>
        | WithModifiers<BodyType>
        | OneOfEntries
        | BodyType;
}

export interface BodyInstance {
    [key: string]: Primitives | BodyInstance | BodyInstance[];
}

export function isBodyType(input: any): input is BodyType {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (entry) =>
            isResourceString(entry) ||
            isBase(entry) ||
            isWithModifiers(entry) ||
            isOneOfEntries(entry) ||
            isBodyType(entry)
    );
}

export interface HeaderMap {
    [key: string]: Exclude<BaseType, null>;
}

export function isHeaderMap(input: any): input is HeaderMap {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (value) => value !== null && isBaseType(value)
    );
}
