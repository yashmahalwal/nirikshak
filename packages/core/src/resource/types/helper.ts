import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import { ResourceBase, isResourceBase } from "./resource";
import { Primitives } from "../../common/types/helpers";

// An instance is of the following type
export interface ResourceInstanceBase {
    [key: string]: Primitives | ResourceInstanceBase | ResourceInstanceBase[];
}

export interface ResourceInstance extends ResourceInstanceBase {
    id: string | number;
}

// Base type: Simple literals that a resource field can have
export type BaseType =
    | FakerType
    | CustomFunctionType
    | Literal
    | Array<BaseType>;

// Validation for the same
export function isBaseType(input: any): input is BaseType {
    if (Array.isArray(input)) return input.every((entry) => isBaseType(entry));

    // Ordering is important
    return isFakerType(input) || isCustomFunction(input) || isLiteral(input);
}

// Higher order type: Adding modifiers to a resource field entry
// A resource base field can be: Base Type or another Resource base
// You can add nullable, plural or optional here
// Ex: {type: "faker:random.words", nullable: true}
// Ex: {field: {name: "custom:myName"}, plural: true}
export type WithModifiers<
    T extends BaseType | ResourceBase
> = (T extends BaseType
    ? {
          type: BaseType;
      }
    : {
          field: ResourceBase;
      }) & { plural?: boolean; nullable?: boolean; optional?: boolean };

// Checking the validity of the modifier fields of the HOT
function checkModifiers(
    input: WithModifiers<ResourceBase | BaseType>
): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersResource(
    input: any
): input is WithModifiers<ResourceBase> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isResourceBase(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBaseType(
    input: any
): input is WithModifiers<BaseType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBaseType(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<BaseType> | WithModifiers<ResourceBase> {
    return isWithModifiersBaseType(input) || isWithModifiersResource(input);
}

// Selecting one from multiple choices
export type OneOfEntries =
    | {
          types: Array<BaseType | WithModifiers<BaseType>>;
      }
    | {
          fields: Array<ResourceBase | WithModifiers<ResourceBase>>;
      }
    | {
          fields: Array<ResourceBase | WithModifiers<ResourceBase>>;
          types: Array<BaseType | WithModifiers<BaseType>>;
      };

//   type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBaseType(entry) || isWithModifiersBaseType(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) =>
                      isResourceBase(entry) || isWithModifiersResource(entry)
              )
            : true);

    return value;
}
