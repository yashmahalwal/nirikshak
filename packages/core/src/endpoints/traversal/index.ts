import Supertest from "supertest";
import { MethodType, Inputs, Outputs } from "../types";
import { URLString } from "../types/urlString";
import { ResourceInstance, Resource } from "../../resource";
import { SchemaHelpers } from "../../common";
import { Collection } from "./collection";
import { HeadersInstance } from "../generation/helpers/headerMapGen";
import { makeGetRequest } from "./get";

export { TraversalHelpers, TraversalHelperFunctions } from "./traversalHelpers";
export { Collection } from "./collection";

export async function makeRequest(
    url: URLString,
    method: MethodType,
    caseValue: keyof Outputs[typeof method],
    server: Supertest.SuperTest<Supertest.Test>,
    input: Inputs[typeof method],
    resource: ResourceInstance,
    helpers: SchemaHelpers,
    collection: Collection,
    resourceJSON: Resource
): Promise<{
    status: number;
    headers?: HeadersInstance;
    body?: any;
}> {

    switch(method){
        case "GET":
            return makeGetRequest(caseValue, server,url,input,resource,helpers,resourceJSON);
    }

    return { status: 200 };
}
