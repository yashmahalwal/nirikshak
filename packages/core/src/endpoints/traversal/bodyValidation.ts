import {
    BodyType,
    BaseType,
    isBaseType,
    Base,
    WithModifiers,
    OneOfEntries,
    isBase,
    isWithModifiersBodyType,
    isWithModifiersBase,
} from "../types/helpers";
import { ResourceInstance, isOneOfEntries } from "../../resource/types/helper";
import { SchemaHelpers, isPrimitives } from "../../common/types/helpers";
import { ResourceString, isResourceString } from "../types/resourceString";
import { getFromResource } from "../generation/resourceStringGen";
import _ from "lodash";
import { isLiteral } from "../../common/types/literals";
import { generateBaseType } from "../generation/helpers/baseTypeGen";
import { isFakerType } from "../../common/types/fakerTypes";
import {
    isCustomFunction,
    CustomFunctionType,
} from "../../common/types/custom";
import { HeaderAndStatus } from "../types";
import {
    TraversalHelpers,
    validateWithTraversalHelpers,
} from "./traversalHelpers";
import { NodeEntry, ParsedNode } from "../graph/nodeTypes";
import { Collection } from "./collection";
import { InputInstance } from ".";

export function matchResourceString(
    input: any,
    resourceString: ResourceString,
    resource: ResourceInstance
): boolean {
    return _.isEqual(input, getFromResource(resourceString, resource));
}

export async function matchBaseType(
    input: any,
    baseType: BaseType,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<boolean> {
    return (
        isLiteral(input) &&
        (isFakerType(baseType) ||
        isCustomFunction(baseType) ||
        isResourceString(baseType)
            ? typeof input ===
              typeof (await generateBaseType(baseType, resource, helpers))
            : input === baseType)
    );
}

export async function matchBase(
    input: any,
    base: Base,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<boolean> {
    if (isBaseType(base)) return matchBaseType(input, base, resource, helpers);
    if (!Array.isArray(input) || input.length !== base.length) return false;

    for (let i = 0; i < base.length; i++)
        if (!(await matchBase(input[i], base[i], resource, helpers)))
            return false;
    return true;
}

export async function matchWithModifiersBase(
    input: any,
    schema: WithModifiers<Base>,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<boolean> {
    // Optional is considered before invoking this function

    if (schema.nullable && input === null) return true;
    if (schema.plural) {
        if (!Array.isArray(input)) return false;
        // Fire up matching for every entry
        const promiseArr: Promise<boolean>[] = input.map((entry) =>
            matchBase(entry, schema.type, resource, helpers)
        );
        // Wait for all to be resolved
        const resolved = await Promise.all(promiseArr);
        // Return the verdict
        return resolved.every((returnValue) => returnValue === true);
    }

    // Neither plural, nor nullable
    return matchBase(input, schema.type, resource, helpers);
}

export async function matchWithModifiersBodyType(
    input: any,
    schema: WithModifiers<BodyType>,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<boolean> {
    // Optional is considered before invoking this function

    if (schema.nullable && input === null) return true;
    if (schema.plural) {
        if (!Array.isArray(input)) return false;
        // Fire up matching for every entry
        const promiseArr: Promise<boolean>[] = input.map((entry) =>
            matchBody(entry, schema.field, resource, helpers)
        );
        // Wait for all to be resolved
        const resolved = await Promise.all(promiseArr);
        // Return the verdict
        return resolved.every((returnValue) => returnValue === true);
    }

    // Neither plural, nor nullable
    return matchBody(input, schema.field, resource, helpers);
}

export async function matchOneOfEntries(
    input: any,
    schema: OneOfEntries,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<boolean> {
    // If an optional entry is provided
    if (input === undefined) {
        if ("types" in schema)
            for (const type of schema.types)
                if (isWithModifiersBase(type) && type.optional) return true;
        if ("fields" in schema)
            for (const field of schema.fields)
                if (isWithModifiersBodyType(field) && field.optional)
                    return true;
        return false;
    }

    // If a null entry is provided
    if (input === null) {
        if ("types" in schema)
            for (const type of schema.types)
                if (
                    (isWithModifiersBase(type) && type.nullable) ||
                    type === null
                )
                    return true;
        if ("fields" in schema)
            for (const field of schema.fields)
                if (isWithModifiersBodyType(field) && field.nullable)
                    return true;
        return false;
    }

    // Input is not null or undefined

    if (isPrimitives(input)) {
        // Primitive can only be generated from schema.types
        if ("types" in schema)
            for (const type of schema.types)
                if (isBase(type)) {
                    if (await matchBase(input, type, resource, helpers))
                        return true;
                } else if (
                    await matchWithModifiersBase(input, type, resource, helpers)
                )
                    return true;
    } else if ("fields" in schema) {
        for (const field of schema.fields) {
            if (isWithModifiersBodyType(field)) {
                if (
                    await matchWithModifiersBodyType(
                        input,
                        field,
                        resource,
                        helpers
                    )
                )
                    return true;
            } else if (await matchBody(input, field, resource, helpers))
                return true;
        }
    }

    return false;
}

export async function matchBody(
    body: any,
    bodySchema: BodyType,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<boolean> {
    if (!body || typeof body !== "object") return false;

    let value = true;
    for (const key in bodySchema) {
        const entry = bodySchema[key];
        if (isResourceString(entry)) {
            if (body[key] === undefined) {
                try {
                    getFromResource(entry, resource);
                    value = false;
                } catch {
                    continue;
                }
            }
            value = matchResourceString(body[key], entry, resource);
            void value;
        } else if (isBase(entry))
            value = await matchBase(body[key], entry, resource, helpers);
        else if (isWithModifiersBase(entry))
            value =
                entry.optional && body[key] === undefined
                    ? true
                    : await matchWithModifiersBase(
                          body[key],
                          entry,
                          resource,
                          helpers
                      );
        else if (isWithModifiersBodyType(entry))
            value =
                entry.optional && body[key] === undefined
                    ? true
                    : await matchWithModifiersBodyType(
                          body[key],
                          entry,
                          resource,
                          helpers
                      );
        else if (isOneOfEntries(entry))
            value = await matchOneOfEntries(
                body[key],
                entry,
                resource,
                helpers
            );
        else value = await matchBody(body[key], entry, resource, helpers);

        if (!value) return false;
    }

    return value;
}

export function extractBodiesFromOutput(
    input: {
        semantics: HeaderAndStatus;
        body?: BodyType | CustomFunctionType;
    }[]
): (
    | { type: "body type"; value: BodyType }
    | { type: "custom function"; value: CustomFunctionType }
)[] {
    const bodyArr: (
        | { type: "body type"; value: BodyType }
        | { type: "custom function"; value: CustomFunctionType }
    )[] = [];
    input.forEach(
        (entry) =>
            "body" in entry &&
            bodyArr.push(
                isCustomFunction(entry.body)
                    ? { type: "custom function", value: entry.body }
                    : { type: "body type", value: entry.body as BodyType }
            )
    );
    return bodyArr;
}

export async function bodyValidation(
    body: any,
    bodyArr: (
        | { type: "body type"; value: BodyType }
        | { type: "custom function"; value: CustomFunctionType }
    )[],
    resource: ResourceInstance,
    schemaHelpers: SchemaHelpers,
    traversalHelpers: TraversalHelpers,
    nodeEntry: NodeEntry,
    collection: Collection,
    input: InputInstance,
    parsedNode: ParsedNode
): Promise<boolean> {
    return (
        !bodyArr.length ||
        (
            await Promise.all(
                bodyArr.map((b) =>
                    b.type === "body type"
                        ? matchBody(body, b.value, resource, schemaHelpers)
                        : validateWithTraversalHelpers(
                              body,
                              b.value,
                              resource,
                              schemaHelpers,
                              traversalHelpers,
                              nodeEntry,
                              collection,
                              input,
                              parsedNode
                          )
                )
            )
        ).some((val) => val)
    );
}
