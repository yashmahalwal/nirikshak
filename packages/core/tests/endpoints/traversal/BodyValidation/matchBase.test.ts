import faker from "faker";
import { ResourceInstance } from "../../../../src/resource/types/helper";
import { SchemaHelpers } from "../../../../src/common/types/helpers";
import { Base } from "../../../../src/endpoints/types/helpers";
import { matchBase } from "../../../../src/endpoints/traversal/bodyValidation";
const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidResource: ResourceInstance = {
    created: true,
    id: faker.random.uuid(),
    args: [1, 2, 3, [false], true, "strstr", null],
    addresses: [{ zipCode: "370432" }, { zipCode: "232344" }],
    fallacies: faker.random.word(),
    places: {
        to: {
            be: {
                miles: {
                    to: {
                        go: {
                            before: 23,
                        },
                    },
                },
            },
        },
    },
};

const Entries: {
    input: any;
    base: Base;
    output: boolean;
}[] = [
    {
        input: "Worder",
        base: "custom:word",
        output: true,
    },
    {
        input: 88,
        base: "number",
        output: false,
    },
    {
        input: 4,
        base: [1, 2, 3, 4],
        output: false,
    },
    {
        input: [1, 2, 3],
        base: [1, 2, 3, 4],
        output: false,
    },
    {
        input: [1, 2, 3, 4],
        base: [1, 2, 3, "faker:random.number"],
        output: true,
    },
    {
        input: [1, 2, 3, ValidResource.id],
        base: [1, 2, 3, "resource:id"],
        output: true,
    },
    {
        input: [1, 2, [[[3], "word1"], "word2"], 4],
        base: [
            1,
            2,
            [[[3], "custom:word"], { function: "faker:random.word", args: [] }],
            "faker:random.number",
        ],
        output: true,
    },
    {
        input: [1, 2, 3, 4],
        base: [1, 2, 3, "faker:random:word"],
        output: false,
    },
    {
        input: [1, 2, [[[3], "word1", [[[[["str"]]]]]], "word2"], 4],
        base: [
            1,
            2,
            [
                [[3], "custom:word", [[[[[null]]]]]],
                { function: "faker:random.word", args: [] },
            ],
            "faker:random.number",
        ],
        output: false,
    },
];

const InvalidEntries: { input: any; base: any }[] = [
    {
        input: true,
        base: "resource:addresses",
    },
    {
        input: false,
        base: "resource:lorem.ipsum",
    },
    {
        input: "worldy",
        base: {
            function: "custom:nonExistentHelper",
        },
    },
    {
        input: "D. Piscatella",
        base: "custom:invalidHelper",
    },
    {
        input: [1, 2, [[[3], "word1", [[[[["str"]]]]]], "word2"], 4],
        base: [
            1,
            2,
            [
                [[3], "custom:invalidHelper", [[[[[null]]]]]],
                { function: "faker:random.word", args: [] },
            ],
            "faker:random.number",
        ],
    },
    {
        input: [1, 2, [[[3], "word1", [[[[[null]]]]]], "word2"], 4],
        base: [
            1,
            2,
            [
                [[3], "custom:word", [[[[[null]]]]]],
                { function: "faker:random.word", args: [] },
            ],
            "custom:nonExistentHelper",
        ],
    },
    {
        input: [1, 2, [[[3], "word1", [[[[[null]]]]]], "word2"], 4],
        base: [
            1,
            2,
            [
                [[3], "resource:addresses", [[[[[null]]]]]],
                { function: "faker:random.word", args: [] },
            ],
            "faker:random.number",
        ],
    },
    {
        input: [1, 2, 3, 4],
        base: [1, 2, "resource:lorem.ipsum", "faker:random.number"],
    },
];

describe(`Base type match`, () => {
    test.each(Entries)(`Valid base entry: %#`, (entry) =>
        expect(
            matchBase(entry.input, entry.base, ValidResource, Helpers)
        ).resolves.toBe(entry.output)
    );

    test.each(InvalidEntries)(`Invalid base entry: %#`, (entry) =>
        expect(
            matchBase(entry.input, entry.base as Base, ValidResource, Helpers)
        ).rejects.toMatchSnapshot()
    );
});
