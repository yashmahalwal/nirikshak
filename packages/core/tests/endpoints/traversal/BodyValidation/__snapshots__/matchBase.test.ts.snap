// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base type match Invalid base entry: 0 1`] = `[Error: Resource string resource:addresses resolved to non primitive value]`;

exports[`Base type match Invalid base entry: 1 1`] = `[Error: Invalid resource instance path: lorem.ipsum]`;

exports[`Base type match Invalid base entry: 2 1`] = `[Error: function: custom:nonExistentHelper, args: ]`;

exports[`Base type match Invalid base entry: 3 1`] = `[Error: Non literal type value from custom:invalidHelper]`;

exports[`Base type match Invalid base entry: 4 1`] = `[Error: Non literal type value from custom:invalidHelper]`;

exports[`Base type match Invalid base entry: 5 1`] = `[Error: function: custom:nonExistentHelper, args: ]`;

exports[`Base type match Invalid base entry: 6 1`] = `[Error: Resource string resource:addresses resolved to non primitive value]`;

exports[`Base type match Invalid base entry: 7 1`] = `[Error: Invalid resource instance path: lorem.ipsum]`;
